cpp/modmesh/buffer/SimpleCollector.hpp

```diff
@@ -0,0 +1,143 @@
+ *
+ * - Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ * - Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ * - Neither the name of the copyright holder nor the names of its contributors
+ *   may be used to endorse or promote products derived from this software
+ *   without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <modmesh/buffer/BufferExpander.hpp>
+#include <modmesh/buffer/SimpleArray.hpp>
+
+#include <limits>
+#include <stdexcept>
+
+namespace modmesh
+{
+
+template <typename T>
+class SimpleCollector
+{
+
+private:
+
+    using array_internal_types = detail::SimpleArrayInternalTypes<T>;
+
+public:
+
+    using value_type = typename array_internal_types::value_type;
+    using expander_type = BufferExpander;
+
+    static constexpr size_t ITEMSIZE = sizeof(value_type);
+
+    explicit SimpleCollector(size_t length)
+        : m_expander(BufferExpander::construct(length * ITEMSIZE))
+    {
+    }
+
+    SimpleCollector()
+        : m_expander(BufferExpander::construct())
+    {
+    }
+
+    SimpleCollector(SimpleCollector const & other)
+        : m_expander(other.m_expander->clone())
+    {
+    }
+
+    SimpleCollector & operator=(SimpleCollector const & other)
+    {
+        if (this != &other)
+        {
+            m_expander = other.m_expander->clone();
+        }
+        return *this;
+    }
+
+    SimpleCollector(SimpleCollector && other)
+        : m_expander(std::move(other.m_expander))
+    {
+    }
+
+    SimpleCollector & operator=(SimpleCollector && other)
+    {
+        if (this != &other)
+        {
+            m_expander = std::move(other.m_expander);
+        }
+        return *this;
+    }
+
+    size_t size() const { return expander().size() / ITEMSIZE; }
+    size_t capacity() const { return expander().capacity() / ITEMSIZE; }
+
+    value_type const & operator[](size_t it) const noexcept { return data(it); }
```

Comment:
Data collecting / accumulating is usually linear.  I think we only need 1D content access.